Support Coq's proof diffs feature: highlight diffs in goals and some error messages

jfehrle opened the pull request:
This PR highlights Coq proof diffs generated in goals and some error messages.  Showing these requires https://github.com/coq/coq/pull/10019.  This PR can be committed without waiting for that change--it will not break anything.  I hope that PR will be in Coq 8.10, which is scheduled for June.  

https://github.com/coq/coq/pull/10020 is needed to show the highlights when undoing proof steps--PG uses the "Show" command, which before this PR was not highlighting proof diffs at all.

Please see the doc in this PR and the Coq documentation at https://coq.inria.fr/distrib/current/refman/proof-engine/proof-handling.html#showing-differences-between-proof-steps.

Here are some screen shots.  The first shows additions only ("on" setting), the second shows the same proof step with the "removed" setting.  The third (barely legible) shows an error message that includes diffs.

![image](https://user-images.githubusercontent.com/1253341/56876522-dce23c80-69fc-11e9-9edc-d7ec1d76db11.png)

![image](https://user-images.githubusercontent.com/1253341/56876542-fc796500-69fc-11e9-9729-92423bbacb3e.png)

![image](https://user-images.githubusercontent.com/1253341/56876553-131fbc00-69fd-11e9-9aaf-19a939ed84e5.png)

Clearly the default highlight colors need some work.  The purple text in the 3rd screen shot is hard to read on the red background.  Also there's some syntax highlighting that uses the same green as the diff highlighting (below), making some characters disappear entirely.  I thought others might have ideas on how best to handle this--whether different defaults for diffs or perhaps changing some of the default colors for syntax highlighting.

![image](https://user-images.githubusercontent.com/1253341/56876807-cb9a2f80-69fe-11e9-9132-2b91190401cd.png)

The unhighlighted text looks like this:

![image](https://user-images.githubusercontent.com/1253341/56876843-0ef49e00-69ff-11e9-8ac7-7941f7996884.png)

The associated PG issue is https://github.com/ProofGeneral/PG/issues/381.

jfehrle commented:
Two more questions:
1. Coq -> Settings -> Diffs only sets the Coq Diffs setting used at startup time.  It doesn't change the setting in mid-session.  Is there a standard way for the user to change it in mid-session?  Should the doc say to enter a Coq command such as `Set Diffs "on".`?
2. When I generated the HTML documentation, the style was quite different from that used on the website at https://proofgeneral.github.io/doc/userman/.  Also the one I generated said it was for version `4.5-git` versus version `4.4.1~pre` on the website.  What should I make of this?  Of course, I'd like to get out updated documentation so users can benefit from the updated feature.

Mine HTML manual looks like this:

![image](https://user-images.githubusercontent.com/1253341/56877180-043b0880-6a01-11e9-8301-86a3d6431edb.png)

jfehrle commented:
Also, FWIW, it appears that in the green-on-green case, the green syntax highlighting suppresses the underlining for the diff highlight.  At least on my system.

monnier commented:
> This PR highlights Coq proof diffs generated in goals and some error
> messages.  Showing these requires
> https://github.com/coq/coq/pull/10019.  This PR can be committed
> without waiting for that change--it will not break anything.  I hope
> that PR will be in Coq 8.10, which is scheduled for June.  

I don't know how the <diff.added> tags are generated in Coq, but if
they're based on purely textual info (rather than based on structural
info before generating the diff), then an alternative is to use
`smerge-refine-regions` (aka `smerge-refine-subst`) to generate the
highlighting on the Emacs side.

Also, I'd recommend you make the coq-diffs-* faces inherit from the
corresponding diff(-refine)?-(added|removed) faces (or even directly
use these faces without bothering to define your own).

ejgallego commented:
> I don't know how the <diff.added> tags are generated in Coq, but if they're based on purely textual info (rather than based on structural info before generating the diff)

Can you be a bit more concrete @monnier ?

There is nothing too special about the tags, other than them being markers, so i guess they would fall in the "purerly textual info" category.

jfehrle commented:
> I don't know how the <diff.added> tags are generated in Coq, but if
they're based on purely textual info (rather than based on structural
info before generating the diff)

You're suggesting generating the diffs in emacs instead?  The Coq diff is not a text-based diff; it takes advantage of additional information.  A purely text-based diff would be less effective for users.  The documentation (see ProofGeneral.texi in this PR) mentions this:

```
The Coq proof diff provides a better user experience than company-coq diff
because it takes advantage of information not readily available outside of
Coq.  For example:

@itemize @bullet
@item diffs are done on a token basis as determined by the Coq lexer rather
  than on a character basis, probably a better match for how people understand
  the output
@item diffs ignore the order of hypotheses
@item tactics that only change the proof view are handled specially, for
  example "swap" after a "split" will show the diffs between before "split"
  and after "swap", which is more useful
@item some error messages have been instrumented to show diffs where it is helpful
@end itemize
```

> Also, I'd recommend you make the coq-diffs-* faces inherit from the
corresponding diff(-refine)?-(added|removed) faces (or even directly
use these faces without bothering to define your own).

Inherit exactly how?  (This is the first bit of emacs lisp I've done and I'm not a frequent user of emacs.)
When I substitute `diff-added` and `diff-removed` as the faces, I don't get any highlight.  Also, by defining my own faces, they appear in the Coq menu, allowing for easy customization--how would the user change the diff faces if they are inherited from elsewhere?

Are you a merger for PG?

cpitclaudel commented:
> The Coq proof diff provides a better user experience than company-coq diff
because it takes advantage of information not readily available outside of
Coq.

This is a bit misleading, I think.

> @item diffs are done on a token basis as determined by the Coq lexer rather
  than on a character basis, probably a better match for how people understand
  the output

I don't understand what this means

> @item diffs ignore the order of hypotheses

So does company-coq, IIRC

> @item tactics that only change the proof view are handled specially, for
  example "swap" after a "split" will show the diffs between before "split"
  and after "swap", which is more useful

That sounds nice

> @item some error messages have been instrumented to show diffs where it is helpful

Company-coq does that too.

Really, I think there's no need to compare. The feature you implemented is superior, and more importantly once Coq does proper structural diffing the infrastructure will stay the same, so it makes complete sense to add support for your diffs to PG; but, if you must compare, then it'd be better compare accurately :)

jfehrle commented:
> Really, I think there's no need to compare.

Of course.  I'll remove the reference to company-coq.  

> > @item diffs are done on a token basis as determined by the Coq lexer rather
> > than on a character basis, probably a better match for how people understand
> > the output
> 
> I don't understand what this means

Coq passes the text to the lexer and then does the diff based on whether the tokens are equal or not.
For example, a character-based diff between "abc" and "axc" would indicate that "a" and "c" are in common and that only the "b"/"x" is a difference.  The token-based difference would highlight all of "abc" and "axc" as a difference.

> > @item some error messages have been instrumented to show diffs where it is helpful
> 
> Company-coq does that too.

It handles the message in the third screen shot?  Nice.

> once Coq does proper structural diffing the infrastructure will stay the same, so it makes complete sense to add support for your diffs to PG

If it ever does.  I spoke to Yves Bertot (via email) when I was designing the diff feature, .  Apparently at one time there was a structural diff routine in Coq, but it was complex and difficult to maintain.  We thought the sweet spot was to use the token-based diff.

monnier commented:
>> I don't know how the <diff.added> tags are generated in Coq, but if
>> they're based on purely textual info (rather than based on structural
>> info before generating the diff)
> You're suggesting generating the diffs in Emacs instead?

Not the diff, no: the highlighting of the diff (tho I don't know if that
makes sense; I haven't seen this feature myself, nor company-coq's diff
for that matter, I'm only talking based on the patch you sent, and my
knowledge of Emacs and its diff-mode).

> The Coq proof diff provides a better user experience than company-coq diff
> because it takes advantage of information not readily available outside of
> Coq.  For example:
>
> @itemize @bullet
> @item diffs are done on a token basis as determined by the Coq lexer rather
>   than on a character basis, probably a better match for how people understand
>   the output

My experience with smerge-refine-region (where I inspired myself from
Ediff's word-level highlighting but played around with various levels
and kinds of granularity) is that it's very hard to argue that one is
better than another here: some will work better in some cases and other
will work better in other cases (e.g. sometimes you just want to see
"TOKEN-UP" changed to "TOKEN-DOWN" but other times you'd want the
highlighting to be finer-grained and only highlight the "UP" and
"DOWN" part of those two tokens that changed).

> @item diffs ignore the order of hypotheses
> @item tactics that only change the proof view are handled specially, for
>   example "swap" after a "split" will show the diffs between before "split"
>   and after "swap", which is more useful

That's great.  It might indeed make it harder to make good use
of smerge-refine-region.

In any case, I only suggested the use of smerge-refine-region since
there was a good chance you didn't know about it and that maybe it
made more sense than changing Coq.

But either way is perfectly fine (tho, obviously, seeing my
smerge-refine-region code used elsewhere would bring me
satisfaction ;-).

>> Also, I'd recommend you make the coq-diffs-* faces inherit from the
>> corresponding diff(-refine)?-(added|removed) faces (or even directly
>> use these faces without bothering to define your own).
> Inherit exactly how?

Beside :background, :foreground, :width, etc... there's also a :inherit
attribute which you can set to a face (or a list of faces).  E.g.:

    (defface coq-diffs-added-face
      '((t :inherit diff-refine-added))
      "Face used to highlight added text in diffs"
      :group 'proof-faces)

> When I substitute `diff-added` and `diff-removed` as the faces, I don't get
> any highlight.

What if you do `M-x describe-face RET diff-added RET`?
Any chance `diff-mode` is simply not loaded (so you need to add
a `(require 'diff-mode)`)?

> how would the user change the diff faces if they are inherited
> from elsewhere?

You can combine the :inherit with other attributes (which are then
added "on top of" of those from which we inherit).

And if you directly use the `diff-added` face, then the user can find it
in "the usual" way: place the cursor on the text whose appearance you
want to change and do `M-x customize-face RET`.

> Are you a merger for PG?

I think technically I have the needed rights for it, but I try to avoid
the non-Free Javascript on the github site and haven't bothered to try
and setup something like ghub.el, so in practice, no.

cpitclaudel commented:
> I think technically I have the needed rights for it, but I try to avoid the non-Free Javascript on the github site and haven't bothered to try and setup something like ghub.el, so in practice, no.

But if Stefan tells me to merge, I will :)

ejgallego commented:
Does the approach of sending `<diff> ... </diff>` tags is ok for you folks?

If so, we will merge the Coq PR adding this to `-emacs`.

jfehrle commented:
> I haven't seen this feature myself

Well, I hope you saw the screenshots.  FWIW, here is the Coq output corresponding to the first screenshot.  But I still don't understand the original comment.  If you wish, please elaborate.

```
<prompt>count_rev_tree < 9 |count_rev_tree| 0 < </prompt>2 subgoals (ID 22)
  T : Type
  p : T -> bool
  ============================
  <diff.added.bg>count p (rev_tree <diff.added>Leaf</diff.added>) = count p <diff.added>Leaf</diff.added></diff.added.bg>

subgoal 2 (ID 29) is:
 <diff.added.bg>count p (rev_tree <diff.added>(Node a t1 t2)</diff.added>) = count p <diff.added>(Node a t1 t2)</diff.added></diff.added.bg>
```

I got the inherit to work.  However, this is not the most legible combination of colors.   I'm using the out-of-the-box color settings.  Should the out-of-the-box settings be different, or do most/all emacs users customize their colors?

![image](https://user-images.githubusercontent.com/1253341/56918740-b27f9600-6a73-11e9-8343-6b2cd5437cc5.png)

The purple string text looks OK, though:

![image](https://user-images.githubusercontent.com/1253341/56919139-b233ca80-6a74-11e9-83c3-819f02b0852f.png)

Using inheritance seems more consistent than using these faces directly, since the `defface` also adds it to the Coq menus with the Coq-defined faces--a logical place to look for them, even IIUC emacs users generally eschew menus.  (Also `customize-face` doesn't seem so easy to discover--though maybe that's just because I'm an infrequent emacs user.)

![image](https://user-images.githubusercontent.com/1253341/56919880-94fffb80-6a76-11e9-8d18-6d12a79e9d46.png)

monnier commented:
> Well, I hope you saw the screenshots.

Now I have.

> FWIW, here is the Coq output corresponding to the first screenshot.  But I still don't understand the original comment.  If you wish, please elaborate.
>
> ```
> <prompt>count_rev_tree < 9 |count_rev_tree| 0 < </prompt>2 subgoals (ID 22)
>   T : Type
>   p : T -> bool
>   ============================
>   <diff.added.bg>count p (rev_tree <diff.added>Leaf</diff.added>) = count p <diff.added>Leaf</diff.added></diff.added.bg>
>
> subgoal 2 (ID 29) is:
>  <diff.added.bg>count p (rev_tree <diff.added>(Node a t1 t2)</diff.added>) = count p <diff.added>(Node a t1 t2)</diff.added></diff.added.bg>
> ```

Indeed, on this output smerge-refine-region is simply inapplicable.

> I got the inherit to work.  However, this is not the most legible
> combination of colors.   I'm using the out-of-the-box color settings.
> Should the out-of-the-box settings be different, or do most/all emacs
> users customize their colors?

It's using the standard color scheme used in most other similar
circumstances/tools (e.g. github diffs and many others), so it's
expected to work well, but of course, it also depends on other faces in
use at the same time.

The colors I see in your screenshot don't match the colors I see here
when I set my Emacs to use white-on-black.  It looks like your Emacs is
using the colors corresponding to black-on-white (aka light background)
for some reason.

Can you try to `M-x customize-variable RET frame-background-mode RET` to
`dark`?

Might be a bug in Emacs (or maybe you're using Emacs within
a text-terminal and just failed to tell Emacs that your text-terminal
uses a dark background?).

erikmd commented:
Dear @jfehrle,
> When I generated the HTML documentation, the style was quite different from that used on the website at https://proofgeneral.github.io/doc/userman/. Also the one I generated said it was for version 4.5-git versus version 4.4.1~pre on the website. What should I make of this? Of course, I'd like to get out updated documentation so users can benefit from the updated feature.

Good point: the online documentation is generated manually from the file [ProofGeneral.texi](https://github.com/ProofGeneral/PG/blob/master/doc/ProofGeneral.texi) with a custom style and a texi2html-based [Makefile](https://github.com/ProofGeneral/proofgeneral.github.io/blob/master/doc/Makefile). Admittedly it'd be useful to automate this further, e.g. via Travis CI!
In the mentime, I'll generate the corresponding online doc as soon as #421 is merged.

jfehrle commented:
@monnier @cpitclaudel The 8.10 release train is moving swiftly.  If one you can confirm that the tagging approach will work for PG in the next day or so, we can get the essential Coq change (https://github.com/coq/coq/pull/10019) into 8.10.  Otherwise it will slip to 8.11, some 6 months from now.

Perhaps your confirmation that the approach is good could be done before we agree on the final form of this PR?

erikmd commented:
AFAIAC I'm very OK with https://github.com/coq/coq/pull/10019, and I don't think objections regarding the tags were raised in this discussion apart from the possible clash with a fancy Coq notation (which is albeit somewhat unlikely).

@Matafou @cpitclaudel @monnier last comment on this before @ejgallego merges the upstream PR?

jfehrle commented:
> Can you try to `M-x customize-variable RET frame-background-mode RET` to
`dark`?

That looks better, perhaps even stylish.  I usually don't mix purple with red nor light green with blue-green. :-)

![image](https://user-images.githubusercontent.com/1253341/56934221-1ae46c80-6aa0-11e9-9495-c36bd771db6e.png)

![image](https://user-images.githubusercontent.com/1253341/56934196-0607d900-6aa0-11e9-8e56-94508d0029ac.png)

> Might be a bug in Emacs (or maybe you're using Emacs within
a text-terminal and just failed to tell Emacs that your text-terminal
uses a dark background?).

I'm running emacs in Windows Subsystem for Linux on Windows.  I'm pretty sure it's just a text terminal.  WSL doesn't support the graphics API.  And it wasn't something I knew to set up.

That leaves this one unanswered question.  Any takers?

> Coq -> Settings -> Diffs only sets the Coq Diffs setting used at startup time. It doesn't change the setting in mid-session. Is there a standard way for the user to change it in mid-session? Should the doc say to enter a Coq command such as Set Diffs "on".?

erikmd commented:
Hi @jfehrle,

> I'm running emacs in Windows Subsystem for Linux on Windows. I'm pretty sure it's just a text terminal. WSL doesn't support the graphics API. And it wasn't something I knew to set up.

Indeed, WSL provides a CLI-only Linux environment (but as I've mentioned recently in another issue in the Coq repo, is possible to run "Coq-WSL" + ProofGeneral from a native Windows-GUI Emacs: this is not yet documented in [the Coq wiki](https://github.com/coq/coq/wiki/Installation-of-Coq-on-Windows), but I've written some related French doc [here](https://github.com/erikmd/tapfa-init.el/))

> Coq -> Settings -> Diffs only sets the Coq Diffs setting used at startup time. It doesn't change the setting in mid-session. Is there a standard way for the user to change it in mid-session? Should the doc say to enter a Coq command such as Set Diffs "on".?

No need to add such a remark in the doc: I've just compiled coq/coq#10020 and ProofGeneral/PG#421 and I confirm the actual behavior is as follows: when you do `Coq > Settings > Diffs > "on"`, the `Set Diffs "on".` command is indeed sent to the proved: you can notice this at the end of the `*coq*` buffer obtained by doing `M-x switch-to-buffer RET *coq* RET`. The only possible interference with this behavior is related to backtracking (if you enable Diffs at some point in a proof, and backtrack the locked region, as `Set Diffs` is a regular vernacular command compatible with backtracking, its effect will be disabled...)

cpitclaudel commented:
> as Set Diffs is a regular vernacular command compatible with backtracking, its effect will be disabled...)

Sounds liek a bug to me; shouldn't Set Diffs be like Set Ltac Debug, immune to backtracking?

erikmd commented:
> shouldn't Set Diffs be like Set Ltac Debug, immune to backtracking?

It could... but I don't see a reason why this would be necessary (I mean, compared to all other options such as `Set Printing All` etc., which are also erased if one backtrack before their definition time.)

jfehrle commented:
> when you do Coq > Settings > Diffs > "on", the Set Diffs "on". command is indeed sent to the prover

I think I got a little confused because I had added `Set Diffs "on"` in my proof script.  After a while it's easy to not see what's actually in the script (or code)!  There are two use cases.  I'd like to support both of them if that's reasonable:
- Set the startup default.  I like to think that many users would want to have diffs turned all the time.
- Change the session setting only.  May be useful on occasion, e.g. to switch from "on" to "removed" when the "on" diffs don't show enough detail.  The user can add the command to the script but it would be nicer if they could leave the script unchanged.

I suppose I could have two menu entries "Diffs" and "Diffs (default)", provided there's an easy way for the code to copy the default setting to the session setting at startup and when the default is changed.  But that seems a bit hokey.  Also users might not expect that changing the default changes the session setting.

If there's not a reasonable/conventional way to support both, I think the first is more important (i.e. the current behavior).

> Sounds like a bug to me; shouldn't Set Diffs be like Set Ltac Debug, immune to backtracking?

Perhaps.  But as @erikmd points out, this is normal behavior for options.  `Set Ltac Debug` is a developer setting.  But it seems very minor, quite a fine point.  I expect most users will decide to have diffs always enabled.  (What would you guys do as users?)

If there are no further review comments, I'll push an updated version with the updated doc.  Probably not today, though.  coq/coq#10019 was merged to master today though it's not yet copied to 8.10.

monnier commented:
> I suppose I could have two menu entries "Diffs" and "Diffs (default)",

I think the other direction makes more sense: don't provide any option
and just always use the highlighting when available (but use the
`font-lock-face` property rather than the `face` property, so the user
can get rid of the colors in the usual way, i.e. by disabling
`font-lock-mode`).


        Stefan

cpitclaudel commented:
> I think the other direction makes more sense: 

I'm not sure. I think our default should mirror the one in CoqIDE, and I don't think it's used by default there.

jfehrle commented:
> I think our default should mirror the one in CoqIDE

CoqIDE lets you select the Diffs option from the menus. The current setting is saved and applied the next time CoqIDE starts up.  The installation default is "off", which was not my preference :-(.

erikmd commented:
BTW regarding the remark about backtracking:
> shouldn't Set Diffs be like **Set Ltac Debug**, immune to backtracking?

it seems the contrary holds… cf. the example session below with a recent Coq master:

```
$ docker run --rm -it coqorg/coq:dev rlwrap coqtop
Welcome to Coq 20af3b9c700f:/home/coq/.opam/4.05.0/.opam-switch/build/coq.dev/_build/default,
master (fcba5e87be13bc5a5374fe274476cd4d5c45f058)

Coq < Goal True.
1 subgoal
  
  ============================
  True

Unnamed_thm < idtac "nodebug".
nodebug
Unnamed_thm < Set Ltac Debug.
Unnamed_thm < idtac "debug".
Goal:
  
  ============================
   True

Going to execute:
idtac "debug"

TcDebug (0) > 

debug

Unnamed_thm < Undo.
Unnamed_thm < Undo.
Unnamed_thm < idtac "nodebug".
nodebug
```

cpitclaudel commented:
> it seems the contrary holds…

Interesting; looks like it changed since (*ahem*) 8.4!

ejgallego commented:
> Interesting; looks like it changed since (_ahem_) 8.4!

It has indeed; per-document options do require using some newer protocol.

erikmd commented:
To follow-up my review comment, I find the current state of the PR for PG master not fully consistent, because:
* If we trigger the menu option `Set Diffs "on".` then backtrack, the option is lost in the current session,
* but reappears at the next session, which is not intuitive (notably as the command menu doesn't give any hint of that state).

On second thought, adding `:pgdynamic t` as I initially suggested is not enough (see below another suggestion).

But I'm sympathetic to the uses cases outlined by @jfehrle:

> I'd like to support both of them if that's reasonable:
> * Set the startup default. […]
> * Change the session setting only. […]
> 
> I suppose I could have two menu entries "Diffs" and "Diffs (default)", […] Also users might not expect that changing the default changes the session setting.

I see two different solutions:

1. Either we document better the current behavior (saying that the menu option `Set Diffs` is kept in all upcoming sessions, albeit possibly not in the current session if one backtracks before the time the menu command has been selected)
2. Or we ensure that the current setting of `Set Diffs` is not preserved across sessions, and add another option `Set Diffs (default)`.

@monnier @cpitclaudel @Matafou do you have a personal preference between these two choices?

Maybe 1. is better for the sake of simplicity?

In any case, I think there are some issues with the current code:

```
(defpacustom diffs "on"
  "Enable diffs display, can be 'on', 'off' or 'removed'"
  :type 'string
  :setting "Set Diffs \"%s\".")
```

(i) It shouldn't be a `defpacustom` but a `defcustom`, in order to ensure that the only possible values for the strings are `on, off, removed`.
I guess you can draw some inspiration from [that setting](https://github.com/ProofGeneral/PG/blob/632a3d7f9ded16faaf58e1c0769bcd4f7c8193e3/coq/coq-compile-common.el#L206-L275) − which is then displayed with a nice menu −, and see also [that comment](https://github.com/ProofGeneral/PG/blob/fb3b75dab55b6e6befffc53e136422558be5faa0/coq/coq.el#L1890-L1894) − notably if the solution 1. above is selected.

(ii) the `Set Diffs` menu option should probably appear in another menu, not in menu Settings (e.g. in menu OPTIONS, typical of session-local Coq parameters), or maybe even better, in the top-level menu, just below `Auto Compilation`...

ejgallego commented:
@erikmd IMVHO there are a few issues with goal display that should be solved by switching to explicit goal display. So PG would use [as done in more modern UI protocols] `Show Goals [Options...]` where in options you would pass current printing mode, sentence to diff again, etc...

erikmd commented:
@ejgallego 
> So PG would use [as done in more modern UI protocols] Show Goals [Options...] where in options you would pass current printing mode, sentence to diff again, etc...

indeed, that makes sense!

Regarding the current PR #421, I guess this is an argument for adopting the "solution 1": rather than duplicating the Set-Diffs setting with two menu options, we should probably keep the Set-Diffs setting persistent in the UI (even if for the moment the display protocol is not yet like you describe and thereby cause a small backtracking issue)

erikmd commented on coq/coq.el:
```diff
@@ -2070,6 +2070,11 @@ See  `coq-fold-hyp'."
+(defpacustom diffs "on"
+  "Enable diffs display, can be 'on', 'off' or 'removed'"
+  :type 'string
+  :setting "Set Diffs \"%s\".")
```
Finally I guess you can ignore this remark on `:pgdynamic t`.

erikmd commented:
@jfehrle regarding the `defpacustom` definition and the menu itself, could you check whether it's feasible to refactor the PR as I suggested in my [previous comment](https://github.com/ProofGeneral/PG/pull/421#issuecomment-488137084)?

ejgallego commented:
by the way @erikmd it seems you were the last person to take care of the async branch, do you think it is abandoned for good or is there a chance to revive it.

I am saying this because if we target SerAPI we could make loads of simplifications on the code [from a quick look], including not having to sync comands, removing all the tricky xml parsing stuff. Not sure if it would be better to start over or reuse the current async work [or even depend on a lsp-style communication package], but I dunno, I thought I'd mention it.

Matafou commented:
True. This problem of syncing options with Coq is an old one. The only way
of doing this properly is to re-set options every time it is needed.

That said I still don’t understand why such options are synchronized in
Coq. Or at least why we can’t have an unsynchronized of them.

P

Le mer. 1 mai 2019 à 00:30, Emilio Jesús Gallego Arias <
notifications@github.com> a écrit :

> @erikmd <https://github.com/erikmd> IMVHO there are a few issues with
> goal display that should be solved by switching to explicit goal display.
> So PG would use [as done in more modern UI protocols] Show Goals
> [Options...] where in options you would pass current printing mode,
> sentence to diff again, etc...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ProofGeneral/PG/pull/421#issuecomment-488139009>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABFSL2SAIVHVFR4REBFMCSLPTDB7TANCNFSM4HJAKFAQ>
> .
>

ejgallego commented:
You can have them "unsynchronized" , not just as a sentence of the proof document. We are not willing to trade incremental document processing as to support them.

jfehrle commented:
> (@monnier) I think the other direction makes more sense: don't provide any option
and just always use the highlighting when available

`Diffs` is a three-valued option, so I think it should be in the menus.

jfehrle commented:
@erikmd The second commit in the version I just pushed does the pacustom refactor, but I didn't know how to have a change in value generate the `Set Diffs "xx"` command.  Can you tell we what I need in the setter?  I put the `Diffs` menu item just below `Auto Compilation` as you suggested.

> 1. Either we document better the current behavior (saying that the menu option Set Diffs is kept in all upcoming sessions, albeit possibly not in the current session if one backtracks before the time the menu command has been selected)
 > 2. Or we ensure that the current setting of Set Diffs is not preserved across sessions, and add another option Set Diffs (default).

The option should be persistent.  I'll update the doc to explain that.  I will also explain the interaction with `Set Diffs` commands in the script.

BTW, I've found typos in other files, too.  I can add their fixes to this PR or a new one.  Let me know if you have a preference.

erikmd commented:
Dear @jfehrle, regarding your questions:

> BTW, I've found typos in other files, too. I can add their fixes to this PR or a new one. Let me know if you have a preference.

As there are several typos, I'd say it's better to fix them in a separate commit (but no need to open another PR just for that).

> is a group required?

Yes, it's better to put specify `:group 'coq` to get a better UX with the command <kbd>M-x customize</kbd>

I've tweaked your code locally without committing, and the following works for setting the option in an active session:

```elisp
(defun coq-diffs--cmd ()
  "Return string for setting Coq Diffs.
Return the empty string if the version of Coq < 8.10."
  (if (coq--post-v810)
      (format "Set Diffs \"%s\". " (symbol-name coq-diffs))
    ""))

(defun coq-diffs--setter (symbol new-value)
  ":set function fo `coq-diffs'.
Set Diffs setting if Coq is running and has a version >= 8.10."
  (set symbol new-value)
  (if (proof-shell-available-p)
      (let ((cmd (coq-diffs--cmd)))
        (if (equal cmd "")
            (message "Ignore coq-diffs setting %s for Coq before 8.10"
                 (symbol-name coq-diffs))
          (proof-shell-invisible-command cmd)))))

(defcustom coq-diffs 'off
  "Controls Coq Diffs option"
  :type '(radio
    (const :tag "Don't show diffs" off)
    (const :tag "Show diffs: only added" on)
    (const :tag "Show diffs: added and removed" removed))
  :safe (lambda (v) (member v '(off on removed)))
  :set 'coq-diffs--setter
  :group 'coq)
```

Note that I also added a `:safe ...` property which can be pretty useful here.

erikmd commented:
There is still something missing to implement:

The code I proposed in my last comment doesn't "hook" into `(proof-assistant-settings-cmds)` so even if the setting is saved from Emacs side, **it is not sent to the proved at startup**.

I see two possible ways to do this:

1. either we fully extend `pacustom` (and related material) to symbol settings (beyond the large refactoring that it represents, an issue that should also be taken into account is that this "special" setting should not be sent to the prover (e.g. at startup) for Coq versions < 8.10)
2. or we just extend `proof-assistant-format` (used by function `proof-assistant-settings-cmds`) to support an additional type, say `%l`, to take a lambda as input (e.g. `(coq-diffs--cmd)`, maybe renamed to `(coq-diffs)`), and add the following code in `coq.el`:

    ```elisp
    (add-to-list proof-assistant-settings
                 '(coq-diffs "%l" string "Show Diffs in Coq") t)
    ```

As `Diffs` is a "special" setting (not boolean) and there are very few similar settings in Coq (actually I couldn't name another such), it seems to me that the **option 2** above could be sufficient for implementing this PR…

@Matafou @cpitclaudel @monnier are you OK with this proposal? or do you have comments?

erikmd commented:
@ejgallego 

> by the way @erikmd it seems you were the last person to take care of the async branch, do you think it is abandoned for good or is there a chance to revive it.

I don't think it is abandoned for good, or at least I hope it isn't…

To give a bit more context, Paul had submitted the pg-xml code for review in #153 and given the size of the PR (716 commits, changing 488 files) and the time constraints (we had to release the code before he leaves), we finally integrated his work as several patches in a new branch, hoping that we could continuously test and improve this `async` branch. This process has somewhat stalled, and unfortunately the branch is not yet usable yet given there are several blocking bugs.
I'm not sure but I suspect a number of these bugs are directly related to artifacts of Coq's XML protocol itself.

Maybe a possible direction would be to create yet another branch `serapi`, reusing features from latest `master`, some code from `async` (e.g. front-end), to ultimately switch the backend protocol from XML to SerAPI?
This is a large amount of work, but I guess it could be even larger if we don't reuse `async`…

Finally, there is a orthogonal issue that should be addressed at the same time: the `async` branch has removed `proof-shell` (which thereby prevents one to run other provers such as Easycrypt), while we may want to have a branch which provides two backends in the same codebase and behaves accordingly.

(BTW @cpitclaudel didn't you implement some time ago a small PG clone with SerAPI? Maybe I'm mistaken, as I wasn't able to find the URL again.)

erikmd commented:
BTW the discussion derived a bit from the initial topic of this PR (implementing Coq Diffs in PG master), so if we need to discuss more about `async` and SerAPI, let's switch to the meta issue #230 instead :)

jfehrle commented:
> coq-diffs--cmd

Why the double dash?

jfehrle commented:
@erikmd I tried your changes.  They set the option in-session, but the setting is not preserved in `~/.emacs`, so it's lost when you exit PG/emacs.  Off topic: after updating Windows last night, I'm seeing a bunch of gross rendering errors, particularly when I'm navigating menus :-(.  But surely not PG related.

> **[the setting] is not sent to the prover at startup.** ... [2 options]

Perhaps a third option is not to show "Diffs" in the menu unless Coq is 8.10 or above.  (I'd go for the simpler solution.)

> this "special" setting should not be sent to the prover (e.g. at startup) for Coq versions < 8.10)

FWIW, the setting exists in 8.9 but the PG user won't see any highlights.

erikmd commented:
@jfehrle
> They set the option in-session, but the setting is not preserved in ~/.emacs, so it's lost when you exit PG/emacs

Normal, because you didn't do `Coq > Settings > Save Settings`

so maybe it'd be good as well to put the `Diffs` menu under `Settings`? but YMMV

erikmd commented:
> coq-diffs--cmd

Why the double dash?

just to indicate that the function is internal, so external libraries should not rely on it

cf. https://www.gnu.org/software/emacs/manual/html_node/elisp/Coding-Conventions.html
> Use two hyphens to separate prefix and name if the symbol is not meant to be used by other packages. (…)

erikmd commented:
but actually i believe it'd make more sense to name it `(defun coq-diffs`  
rather than `(defun coq-diffs-cmd` ...

jfehrle commented:
> so maybe it'd be better to put the Diffs menu under Settings? but YMMV

I mention `Coq > Settings > Save Settings` in the manual.  Everyone will read the manual, right?

Off topic: I notice that, unlike the Coq repository, I can't Edit others' comments.  Not that I want to change them, but it's a convenient way to copy markup symbols.

I pushed my latest changes, I think that only leaves sending the option at startup time.

erikmd commented:
Hi @jfehrle,

> I mention Coq > Settings > Save Settings in the manual. Everyone will read the manual, right?

I hope so. Not only the emacs manual will be automatically available in the Emacs/MELPA package itself after the merge in master, cf. <kbd>C-h i g (ProofGeneral) RET</kbd>, but I'll take care of updating the online doc (and maybe look at ways to automate this soon).

> I pushed my latest changes,

Very good! thanks!

> I think that only leaves sending the option at startup time.

I've just pushed a commit in your branch that should address this (I followed idea 2. from [this comment](https://github.com/ProofGeneral/PG/pull/421#issuecomment-488748985))

@jfehrle please test if this works for you;

@ProofGeneral/core could one of you have a look on the current state of this PR? I believe it is ready, but I wouldn't want to proceed without a green light from your side.

Have a nice WE,
Erik

jfehrle commented:
@erikmd, your change works fine for me (applying the Diffs setting at startup).

The interaction between the Diffs setting and undoing a proof step is strange.  (I think you or someone mentioned this earlier, now I'm paying closer attention.)  For example:

```
(Diffs are off)
tactic1
tactic2
set diffs on
undo
```

shows the tactic1 proof state with Diffs off, yet the PG menu says diffs are on.  Setting Diffs should behave as a global setting rather than as a command inserted in the middle of the proof script, and of course the PG menu should be accurate.  CoqIDE gets around this by sending the settings before every proof step.  If you have a "Set Diffs" in your script, CoqIDE displays the message "Set this option from the IDE menu instead".   How hard would it be to add one or both of these?  This approach may be good for some of the other settings, too.

erikmd commented:
Hi @jfehrle,

> shows the tactic1 proof state with Diffs off, yet the PG menu says diffs are on.

I know, this is the behavior entailed by Coq's semantics of the `Set Diffs "..."` vernacular.

> Setting Diffs should behave as a global setting rather than as a command inserted in the middle of the proof script, and of course the PG menu should be accurate.

It should, but I only see three cases:

1. Either the backtracking semantics of `Set Diffs "…"` is changed in Coq
2. Either we stay as is, and document this fact in the manual, saying e.g. `if you enable Diffs at some point in a proof, and backtrack the locked region, given `Set Diffs` is a regular vernacular command compatible with backtracking, its effect will be disabled, so in this case you may need to click again on the Diffs menu`
3. Either we implement some workaround.

I guess 1. is unlikely to be implemented.
I have no definite option between 2. and 3.

> CoqIDE gets around this by sending the settings before every proof step. If you have a "Set Diffs" in your script, CoqIDE displays the message "Set this option from the IDE menu instead".

Yes, because CoqIDE precisely relies on a different protocol than PG master (which is basically a REPL-based front-end).

> How hard would it be to add one or both of these? This approach may be good for some of the other settings, too.

For the option 3 (workaround), I'd say it would be quite unefficient to send many times further the `Set Diffs` command before any proof step. However, as long as PG is based on the REPL protocol, I can think of another workaround to reduce a bit the "user surprise": after each backtracking (also known as "undo" in the PG terminology), we could maybe test if the current running Coq is >= 8.10, and if yes, trigger the command `Test Diffs` to update the menu accordingly?

Matafou commented:
It is possible to settle something reasonable by remembering where the
option has been set for the last time: each time scripting goes above this
point we need to re-set the option (and remember this point as the new last
time).

P

Le sam. 4 mai 2019 à 23:04, Erik Martin-Dorel <notifications@github.com> a
écrit :

> Hi @jfehrle <https://github.com/jfehrle>,
>
> shows the tactic1 proof state with Diffs off, yet the PG menu says diffs
> are on.
>
> I know, this is the behavior entailed by Coq's semantics of the Set Diffs
> "..." vernacular.
>
> Setting Diffs should behave as a global setting rather than as a command
> inserted in the middle of the proof script, and of course the PG menu
> should be accurate.
>
> It should, but I only see three cases:
>
>    1. Either the backtracking semantics of Set Diffs "…" is changed in Coq
>    2. Either we stay as is, and document this fact in the manual, saying
>    e.g. if you enable Diffs at some point in a proof, and backtrack the
>    locked region, given Set Diffs is a regular vernacular command
>    compatible with backtracking, its effect will be disabled, so in this case
>    you may need to click again on the Diffs menu
>    3. Either we implement some workaround.
>
> I guess 1. is unlikely to be implemented.
> I have no definite option between 2. and 3.
>
> CoqIDE gets around this by sending the settings before every proof step.
> If you have a "Set Diffs" in your script, CoqIDE displays the message "Set
> this option from the IDE menu instead".
>
> Yes, because CoqIDE precisely relies on a different protocol than PG
> master (which is basically a REPL-based front-end).
>
> How hard would it be to add one or both of these? This approach may be
> good for some of the other settings, too.
>
> For the option 3 (workaround), I'd say it would be quite unefficient to
> send many times further the Set Diffs command before any proof step.
> However, as long as PG is based on the REPL protocol, I can think of
> another workaround to reduce a bit the "user surprise": after each
> backtracking (also known as "undo" in the PG terminology), we could maybe
> test if the current running Coq is >= 8.10, and if yes, trigger the command Test
> Diffs to update the menu accordingly?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ProofGeneral/PG/pull/421#issuecomment-489365444>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABFSL2VBI7GPAZHAE576RSTPTX24ZANCNFSM4HJAKFAQ>
> .
>

jfehrle commented:
> > CoqIDE gets around this by sending the settings before every proof step.

> Yes, because CoqIDE precisely relies on a different protocol than PG master (PG using basically a REPL-based protocol).

CoqIde sends a `SetOptions` before every proof step.  The protocol difference doesn't matter, eh?

> I'd say it would be quite unefficient to send many times further the `Set Diffs` command before any proof step.

I expect the difference in performance even if we set Diffs on every command would be imperceptible even when replaying an entire proof script.

Setting the option in the proof script does seem a bit odd, especially doing that more than once (unless you don't know or don't remember the `Diffs` in the menu).  I would lean towards detecting `Set Diffs` commands in the script, printing a message and ignoring them.  That's very simple to explain in the doc and the message makes the behavior easy to discover.  The user won't remember at what point they enabled Diffs from the menu and they may perceive more complex behavior as flaky.  I wouldn't do the `Test Diffs`.

IMHO Coq should at some point make Diffs a global setting not subject to backtracking.

jfehrle commented:
@erikmd, where do we stand on this last issue?  Can we get agreement on an approach upfront?  Are you perhaps working on a solution?  If not, I could look at that--though it would help if you could tell me where to the make the change?

I hope we can complete this PR before Coq 8.10 ships (beginning of June) so that Coq's release notes can give users a simple story for starting to use diffs (i.e., update your PG and try it).

erikmd commented:
> Can we get agreement on an approach upfront?

All in all, it seems the best solution/workaround for PG master is to implement what @Matafou suggested:

> It is possible to settle something reasonable by remembering where the option has been set for the last time: each time scripting goes above this point we need to re-set the option (and remember this point as the new last time).

I'm afraid I won't have time to look at this before next week-end, but maybe Pierre could implement his suggestion, or just suggest some guidelines to help you prepare one such commit?

@jfehrle regarding your other suggestion:

> I would lean towards detecting `Set Diffs` commands in the script, printing a message and ignoring them.

this additional feature could indeed be implemented to make the menu feature easier to discover, but I'd suggest not to ignore the command itself (i.e., just printing a message should be enough...)

> I hope we can complete this PR before Coq 8.10 ships (beginning of June) so that Coq's release notes can give users a simple story for starting to use diffs (i.e., update your PG and try it).

Sounds good :+1:

jfehrle commented:
I think @Matafou's suggestion is fine when there are no `Set Diffs` in the script.  In this case, it's sufficient to unconditionally apply PG's Diffs setting after each `Backtrack`.  I think this is the most important use case.  (It would be odd to add the `Set Diffs` in your script when it's on the menu--unless you don't know it's on the menu.)

If you have a `Set Diffs` in the script and you backtrack across it, do you use Coq's setting (from `Test Diffs`) or PG's setting?  If you use Coq's setting, that's the **opposite** of what PG would do when there are no `Set Diffs` in the script (previous paragraph).  (Also, I think PG's Diffs setting and Coq's Diffs setting must be kept consistent to avoid confusion.)

If you apply PG's Diffs setting when backtracking, then you're not backtracking the effect of `Set Diffs`.  It would be odd to apply `Set Diffs` when going forward in the script but not to undo them when going backward.  So I still think we should detect `Set Diffs` in the script, print a message and ignore them.  That is the simplest to explain.

Your suggested doc text `if you enable Diffs at some point in a proof, and backtrack the locked region, given "Set Diffs" is a regular vernacular command compatible with backtracking, its effect will be disabled, so in this case you may need to click again on the Diffs menu` is complex and sounds like an unfixed bug.  Whereas `"Set Diffs" commands in the script are not applied.  Select the Diffs setting from the menu.` is short and easily understood.  This is the model CoqIDE uses.  The backtrack would always apply the PG Diffs setting.

What do you think?  Can you live with this approach?  There is no perfect solution.

Matafou commented:
Again we must allow people to keep on using set commands inside scripts. A lot of people still set them this way and I see no good reason to make this impossible. It is actually pretty useful sometimes, for instance I often put a Set Debug at some place to trigger ltac debugging precisely (and repeatedly) at some particular line each time I reach it. You are right it does not behave well with menus but a lot of users don’t care. It is also very useful in tutorials.

The obvious solution is to have a “strict synchronized option discipline” option to trigger your idea. If the option is unchecked I suggest we maintain the current behavior. Unless someone has a better idea.

Matafou commented:
One more point: all synchronized options currently have the same issue: if you mix script commands and menu option setting it is the latter is basically broken. I don't think  lot of people ever complained about it, mostly because it is quite uncommon to mix the two styles.

jfehrle commented:
`Set Diffs` and `Set Debug` would be used differently.  I would not ignore `Set Debug` in scripts.  But creating a “strict synchronized option discipline” seems like overkill for a small issue.  Based on your last comment, I won't try to make script commands and menu option settings play together perfectly.  I think that leaves 2 simple alternatives:

- After a Backtrack, always apply the menu setting for Diffs.  (For scripts with `Set Diffs`, this will sometimes change the effective value in Coq.)  Don't worry about consistency between the menu's and Coq's Diffs setting.
- Don't change the code.  Possibly mention that there are quirks in the documentation.

I would lean toward the first.  Is that agreeable?

Matafou commented:
Looks good to me.
We shall deal with option synching either later or in a future implemenation with the newer protocols.

jfehrle commented:
I pushed an updated version.  If it's agreeable, please approve.  Then I'll squash my commits (except the typos) before the merge.  @erikmd, should I squash either of your commits?  Thanks.

erikmd commented:
Thanks @jfehrle! I did not compiled/tested the latest version but the patch LGTM.

> Then I'll squash my commits (except the typos) before the merge. @erikmd, should I squash either of your commits?

I don't think so, I believe my two commits should be semantically kept apart.
YMMV, but I'd even say that your 4 commits need not be squashed, as the Git history of the branch is reasonably clean.

So I guess the PR is ready; since I committed twice in this PR I'd say another co-maintainer should approve the PR (e.g. with the GitHub feature), before we can merge.

jfehrle commented:
OK, I'll keep all the commits as they are.

erikmd commented on coq/coq-diffs.el:
```diff
@@ -0,0 +1,48 @@
+(require 'coq-db)
+
+(defun insert-with-face (str face)
```
@jfehrle 

Regarding the structure of your PR, given you create a new file `coq/coq-diffs.el` with these functions, I'd suggest:

* Running `M-x checkdoc` in the corresponding Elisp buffer to add missing sections (`Commentary:` etc.)
* Prefixing the new functions with `coq-` or so to avoid possible clashes with other packages

jfehrle commented:
Done.  I added author and copyright info, too.  Should I have a License line, too?  (Only 73 out of 102 files have License.)

erikmd commented:
Hi @jfehrle,
> I added author and copyright info, too.

Thanks!

> Should I have a License line, too?

You can; but as you point out not all files explicitly mention it, basically all PG master is under GPLv2 for the moment, and contributions are expected to be submitted under BSD-3, cf.
https://github.com/ProofGeneral/PG#contributing

Finally if you are OK with doing some squash, you could maybe combine your latest commit with your first commit (or with your first two commits)?

monnier commented:
> You can; but as you point out not all files explicitly mention it, basically
> all PG master is under GPLv2 for the moment, and contributions are expected
> to be submitted under BSD-3, cf.

Please try and avoid GPLv2 and make it at least GPLv2+ (BSD-3-Clause and
GPLv3 are fine as well), otherwise it's technically not compatible with
Emacs's GPLv3.

erikmd commented:
@monnier thanks for your comment, the suggestion was just to keep BSD-3 for Jim's contribution; anyway it'd be good indeed to migrate all PG to GPLv3+, but this is beyond the scope of this PR :)

erikmd commented:
BTW @monnier are you globally OK with the current version of the code in this PR? I won't merge before your (or @Matafou/@cpitclaudel's) approval.

jfehrle commented:
Would you give me the exact text for the license line?  I don't really care what flavor it is.

erikmd commented:
I believe a one-liner like this would be sufficient:
```
;; License:  BSD-3 (3-Clause BSD License)
```

erikmd commented:
Finally (sorry to raise this remark only now), you may want to add a line in the CHANGES file? here:
https://github.com/ProofGeneral/PG/blob/master/CHANGES#L28
before doing your squash

monnier commented:
> BTW @monnier are you globally OK with the current version of the code in
> this PR? I won't merge before your (or @Matafou/@cpitclaudel's) approval.

I have not had time to look at the actual code.  But based on the
discussions, I don't see anything that would cause problem from my point
of view.  Also, I'll point out that while I'm familiar with the coq-mode
code, I'm not familiar with the PG code itself (the part which
interacts with external processes).

jfehrle commented:
Updated and squashed.

jfehrle commented:
I hope someone can do the final review and merge soon.

Matafou commented:
Merged, thanks a lot for this @jfehrle !

jfehrle commented:
My pleasure.  Thanks to the reviewers and especially @erikmd.

When can we put this out in official released version?  If that can happen within the next week, I should be able to update the Coq doc for 8.10 to mention the PG support.

IIUC users are using master rather than the async branch.  I can look at adding diff support in async also, though probably not right away.  Is there anything in particular I should know about that branch up front?

erikmd commented:
Thanks to you Jim for your contribution!

> When can we put this out in an official released version?

The MELPA package is up-to-date and the official way to distribute PG:
https://melpa.org/#/proof-general

But admittedly it is not a release *stricto sensu*.

> If that can happen within the next week, I should be able to update the Coq doc for 8.10 to mention the PG support.

First, I will update the online PG doc this week (before Friday 31 May)
Then, we could plan to make a release (i.e. a new git tag for PG) in June, but beforehand I've a couple of things that I'd like to address − especially the license migration that was mentioned here.

> IIUC users are using master rather than the async branch. I can look at adding diff support in async also, though probably not right away. Is there anything in particular I should know about that branch up front?

As it was said before, the async branch is not yet in a usable state, so not stable enough to consider porting #421 to async for the moment AFAICT. Moreover, #421 relies on the `proof-shell` component of PG which has been removed from async…

jfehrle commented:
@erikmd FWIW, I just noticed that the online doc at https://proofgeneral.github.io/doc/userman/ was last updated in December 2016.  Maybe you knew that.

erikmd commented:
Dear @jfehrle, thanks for the reminder; yes I was aware of what you mention.
Actually I'd like to automate the deployment of the doc within PG's CI so the doc stays always up-to-date… but this is not yet done. Anyway, as promised I have updated the online doc (manually for the moment), and the URL of the page documenting your feature is:
https://proofgeneral.github.io/doc/userman/ProofGeneral_12/#Showing-Proof-Diffs

erikmd commented:
Hi all, FYI I've done a major refactoring of the generation of the PG user manual, so the URLs are more "consistent" (while up to now, the URL of the chapter `11` was named `.../ProofGeneral_12/...`)

### Beforehand

![2019-06-01_23-55-32_Screenshot_PG_userman_beforehand](https://user-images.githubusercontent.com/10367254/58754321-9c7b5000-84cd-11e9-8fa8-ef5a99b79574.png)

### From now on

![2019-06-02_00-28-02_Screenshot_PG_userman_now](https://user-images.githubusercontent.com/10367254/58754322-a43af480-84cd-11e9-801c-e4d3c8fa1b6d.png)

*****

Hence this more stable URL for the documentation of the `Coq Diffs` feature:

https://proofgeneral.github.io/doc/userman/Coq-Proof-General/#Showing-Proof-Diffs

Matafou commented:
Thanks Erik!
